src/System.IO.Pipelines.Compression/CompressionPipelineExtensions.cs

```diff
@@ -121,33 +121,20 @@ public WritableDeflateTransform(CompressionLevel compressionLevel, int bits)
 
                     unsafe
                     {
-                        // TODO: Pin pointer if not pinned
-                        void* inPointer;
-                        if (memory.TryGetPointer(out inPointer))
-                        {
-                            _deflater.SetInput((IntPtr)inPointer, memory.Length);
-                        }
-                        else
-                        {
-                            throw new InvalidOperationException("Pointer needs to be pinned");
-                        }
+                        var handle = memory.GetPinnedMemoryHandle();
+                        _deflater.SetInput((IntPtr)handle.PinnedPointer, memory.Length);
+                        handle.Free();
```

Comment:
Ok. I am surprised to see that we are trying to make interop safe - that is lost cause because of it is inherently unsafe. But we gave up on safety for regular code where it would actually matter...